Support for directives (1x/node atm)

joeflack4 opened the pull request:
Changes:
1.  Added support for basic directives, i.e. ones without arguments, such as Apollo's `@client` directive.
2. Refactor of filterNonConfigFields(), which now utilizes external array configFields, which has also been exported for use by other modules (i.e. tests).
3. tslint update: object-literal-sort-keys: `[false]` --> `[true, "as-needed"]`

dupski commented on src/extensions/apollo/index.ts:
```diff
@@ -0,0 +1,21 @@
+    const querySansTypenames = gqlQuery.replace(/__typename /g, '');
+    const queryWithTarget = addApolloTargetToGqlQueryString(querySansTypenames, target);
+    return 'query { ' + queryWithTarget + ' }';
+}
```
Rather than doing string replacement, I would prefer we embed this information in the query object itself.

I've created an issue to track the implementation of "Directives" which is basically what we're doing: https://github.com/dupski/json-to-graphql-query/issues/7

Looking at the usage of `@client` from https://github.com/apollographql/apollo-link-state

How about something like this:

```ts
const query = {
  mutation: {
    updateNetworkStatus: {
      __args: {
        isConnected: true
      },
      __directives: {
        client: true
      }
    }
  }
};
```

This would then result in:

```graphql
mutation {
    updateNetworkStatus(isConnected: true) @client
}
```

That will work for mutations (or rather nodes where sub-nodes are selected), but in the case of fields, we'll need something slightly different. Maybe:

```ts
const query = {
  Posts: {
    id: true,
    name: true,
    author: new WithDirectives({
      client: true
    })
  }
};
```

Which will give us:

```graphql
query {
  Posts {
    id
    name
    author @client
  }
}
```

If you have a look at how we've implemented `__alias` and `EnumType` in the code, that should give you some pointers on how to implement this :)

joeflack4 replied:
Got it. I'm on it~

joeflack4 replied:
Actually, I have some thoughts on this. I think I'll continue the conversation in issue #7 though.

dupski commented:
Also, I just merged some changes from another contributor which has resulted in some merge conflicts - you just need a couple of extra imports (for EnumType and VariableType) in your index.ts file.

dupski commented on src/extensions/apollo/index.ts:
```diff
@@ -0,0 +1,21 @@
+
+export function objToApolloQuery(object: object, target: string): string {
+    const gqlQuery = jsonToGraphQLQuery(object);
+    const querySansTypenames = gqlQuery.replace(/__typename /g, '');
```
Another way to get rid of all __typename keys, would be to add it to the list of objects keys to ignore, here:

https://github.com/dupski/json-to-graphql-query/blob/master/src/jsonToGraphQLQuery.ts#L51

(all those conditions on that line should probably be added to an array - it feels like that would make a good library 'option' too - "keysToRemove", in case there are other keys that people want stripped)

joeflack4 replied:
Regarding __typename, I concur with both ideas:
- Short term: Adding "__typename" to list of keys to ignore
- Long term: Also think a keysToRemove option.

Regarding the merge conflict, np, I'll fetch and resolve it.

joeflack4 commented on src/extensions/apollo/__tests__/apollo.tests.ts:
```diff
@@ -0,0 +1,71 @@
+                title: 'Diet',
+                id: 'diet',
+                options: {
+                    '__typename': 'diet-options',	// Why does my IDE tell me only this line needs quotation?
```
Hey Russell,

Ah, I see. Would you be able to take a look at my setup from these screenshots here and let me know if I've done something wrong? I'm still getting the lint error message showing up.

Here are some screenshots from VS Code. 
<img width="870" alt="screen shot 2018-06-10 at 12 53 46 pm" src="https://user-images.githubusercontent.com/13045020/41204075-e4cf5574-6cad-11e8-9d20-1f1ed5911742.png">
<img width="588" alt="screen shot 2018-06-10 at 12 54 10 pm" src="https://user-images.githubusercontent.com/13045020/41204076-e4db9a5a-6cad-11e8-8f76-1a4377a2ccc2.png">

Some questions:
1) I edited the tslint.json at the root of the project? Is this the correct location to make this configuration edit?
2) Is there a global tsconfig.json and tslint.json that I should be using? I google around and looked around my computer, but couldn't find.
3) Why would the error message be showing up in the first place? Why does it think '__typename' should be quoted? And only in that one place, but not others? Is this a bug in the linter or something that I don't understand?

joeflack4 commented:
@Russell In regards to linting, I've encountered a new issue.

I've made the following refactor:

```
function filterNonConfigFields(fieldName: string) {
    return fieldName !== '__args' && fieldName !== '__alias' && fieldName !== '__variables'
    && fieldName !== '__typename';
}
```
----->
```
const configFields = ['__args', '__alias', '__variables', '__typename'];

function filterNonConfigFields(fieldName: string) {
    return !configFields.includes(fieldName);
}
```

However, my linter is telling me:
"Property 'includes' does not exist on type string[]"

I found this brought up as an issue in TSLint:
https://github.com/Microsoft/TypeScript/issues/11890

The issue can be solved by updating the "lib" key in the tsconfig.json:
```
"lib": [
  "dom",
  "es7"
],
```

But I don't know if this is a change you want to make, or if you like this kind of refactor. Let me know what you think~

dupski commented:
Hey @joeflack4 

I think it may be better to use String.indexOf() since includes() is not supported in microsoft internet exploder unfortunately and it would be good to maintain compatibility if possible.

Chur,
Russ

dupski commented:
Sorry, Array.indexOf()

joeflack4 commented:
Ahh I see. Hmm, too bad; in this case, the meaning of the code while using "indexOf" is a little less clear. Good at least that VSCode is encouraging browser support.

joeflack4 commented:
Hey Russell~. Today I did a large refactor from what I had last time. Per my latest commit message:
>     1) Refactored Apollo Client's @client directive support into general support for directives. Initially, this supports a single directive with no additional arguments per key in an object.
>     2) In tandem with this refactor, moved code into core jsonToGraphQLQuery file and tests file, away from the extensions/apollo directory structure.
>     3) Refactor of filterNonConfigFields(), which now utilizes external array configFields, which has also been exported for use by other modules (i.e. tests).

dupski commented:
Hey @joeflack4 

I think we'll be fine to support argumentless directives for now. I just spent some time merging your branch in - I made a few tweaks based on most recent master. Unfortunately you can't currently have a query with args AND directives at the moment. I've pushed up a branch https://github.com/dupski/json-to-graphql-query/tree/f-directives "f-directives" which has a failing test to this effect - it would be good if you could pull that into your branch.

I'm actually on the road this week, and I'm not sure if I'll be able to look at this again until next week unfortunately. I'm going to release a new version of the library with the `ignoreFields` option added (this change is also in the branch linked above).

Catch up with you next week!

joeflack4 commented:
Thanks for taking an active role in this merge!

Latest updates pushed:
- Added support for queries that have arguments.
- Added capability for nodes to have args and directives simultaneously.

joeflack4 commented:
@dupski Hey Russell. Just checking in on the status of this PR~

dupski commented:
Hey @joeflack4 , sorry for the delay. Started a new contract this week so had to make sure I had everything ready for that. I will do me best to get this sorted this weekend :)

joeflack4 commented:
Sounds good d(^_^)

dupski commented:
Released in `v1.6.0` @joeflack4 - thanks for your contribution! :)
